crates/storage/storage-api/src/state.rs

```diff
@@ -168,6 +177,19 @@ pub trait StateProviderFactory: BlockIdReader + Send + Sync {
     fn pending_state_by_hash(&self, block_hash: B256) -> ProviderResult<Option<StateProviderBox>>;
 }
 
+/// Trait that provides the hashed state from various sources.
+#[auto_impl::auto_impl(&, Box, Arc)]
+pub trait HashedPostStateProvider {
+    /// Returns the `HashedPostState` of the provided `BundleState`.
+    fn hashed_post_state_from_bundle_state(&self, bundle_state: &BundleState) -> HashedPostState;
```

Comment:
> I think we want to minimize the number of places that we introduce the StateCommitment generic to gain access to its associated types.

yeah, if you ask me, the BundleState type is a bit horrific so I don't really want to add any additional logic in there

Reply:
I still don't really understand why `hashed_post_state_from_bundle_state` requires anything more than `BundleState`, ie why it needs to be stateful. `hashed_post_state_from_reverts` makes more sense to require `&self`. Is there an implementation of this trait that makes use of the actual `self` for that method? Otherwise we shouldn't need this

Reply:
I have added `HashedPostStateProvider` to the `StateProvider` super trait. As such all `StateProviders` must implement `HashedPostStateProvider` (I still need to asses if is a hard requirement but logically I think it makes sense).  For the case of `MemoryOverlayStateProvider` I wanted to avoid making it generic over `StateCommitment` and instead leverage it's field `pub(crate) historical: Box<dyn StateProvider>` which implements `HashedPostStateProvider`. As such the implementation of `HashedPostStateProvider` for `MemoryOverlayStateProvider` is defined using this delegation which requires `&self`. I can certainly look into ways we can make this stateless such as removing `HashePostStateProvider` from the `StateProvider` super trait or adding a `StateCommitment` generic to `MemoryOvelayStateProvider`.

Reply:
I think for those implementations, if it were stateless, you could do `<Type as HashedPostStateProvider>::hashed_post_state_from_bundle_state(bundle_state)` if it were stateless. In the macro it could be `<$historical_type as ...>`. The point is that none of these actually require any state, so it does not make sense to be `&self`